// element-plus/lib/utils/vue/install.js
'use strict';

var shared = require('@vue/shared');
var lodashUnified = require('lodash-unified');

const withPropsDefaultsSetter = (target) => {
  const _p = target.props;
  const props = shared.isArray(_p) ? lodashUnified.fromPairs(_p.map((key) => [key, {}])) : _p;
  target.setPropsDefaults = (defaults) => {
    if (!props) {
      return;
    }
    for (const [key, value] of Object.entries(defaults)) {
      const prop = props[key];
      if (!shared.hasOwn(props, key)) {
        continue;
      }
      if (lodashUnified.isPlainObject(prop)) {
        props[key] = {
          ...prop,
          default: value
        };
        continue;
      }
      props[key] = {
        type: prop,
        default: value
      };
    }
    target.props = props;
  };
};
const withInstall = (main, extra) => {
  main.install = (app) => {
    for (const comp of [main, ...Object.values(extra != null ? extra : {})]) {
      app.component(comp.name, comp);
    }
  };
  if (extra) {
    for (const [key, comp] of Object.entries(extra)) {
      main[key] = comp;
    }
  }
  withPropsDefaultsSetter(main);
  return main;
};
const withInstallFunction = (fn, name) => {
  fn.install = (app) => {
    fn._context = app._context;
    app.config.globalProperties[name] = fn;
  };
  return fn;
};
const withInstallDirective = (directive, name) => {
  directive.install = (app) => {
    app.directive(name, directive);
  };
  return directive;
};
const withNoopInstall = (component) => {
  component.install = shared.NOOP;
  withPropsDefaultsSetter(component);
  return component;
};

exports.withInstall = withInstall;
exports.withInstallDirective = withInstallDirective;
exports.withInstallFunction = withInstallFunction;
exports.withNoopInstall = withNoopInstall;
exports.withPropsDefaultsSetter = withPropsDefaultsSetter;
